Documentation/raspi-spi.md

```diff
@@ -0,0 +1,135 @@
+# Enabling SPI on Raspberry Pi
+
+In most of the cases, it is easy and straight forward to enable SPI for your Raspberry Pi. The basic case can be [found here](https://www.raspberrypi-spy.co.uk/2014/08/enabling-the-spi-interface-on-the-raspberry-pi/).
+
+This page will explain how to setup any SPI. Please refer to the [Raspberry Pi documentation](https://www.raspberrypi.org/documentation/hardware/raspberrypi/spi/) to understand the different SPI available. You should be aware as well that for Raspberry Pi4, some of the configurations are different than for the other version especially for SPI3, 4 and 5. 
+
+## Basic Hardware SPI usage
+
+The most simplest code you can build to use SPI is the following:
+
+```csharp
+using System;
+using System.Device.Spi;
+
+namespace TestTest
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            SpiDevice spi = SpiDevice.Create(new SpiConnectionSettings(0));
+            spi.WriteByte(0x42);
+            var incoming = spi.ReadByte();
+        }
+    }
+}
+```
+
+This will open SPI0, with the default Chip Select. It will then write a byte to the MOSI pin and read 1 byte from the MISO pin. 
+
+If you get something like this, it means you need to check the next sections to activate your SPI0:
+
+```
+Unhandled exception. System.IO.IOException: Error 2. Can not open SPI device file '/dev/spidev0.0'.
+   at System.Device.Spi.UnixSpiDevice.Initialize()
+   at System.Device.Spi.UnixSpiDevice.WriteByte(Byte value)
+   at SpiTest.Program.Main(String[] args) in C:\tmp\TestTest\SpiTest\Program.cs:line 11
+Aborted
+```
+
+
+## Enabling SPI0 without Hardware Chip Select
+
+In very short, this is the line you'll need to add into the `/boot/config.txt` file:
+
+```bash
+sudo nano /boot/config.txt
+```
+
+Add the line: 
+
+```
+dtparam=spi=on
+```
+
+Save the file with `ctrl + x` then `Y` then `enter`
+
+Then reboot:
+
+```bash
+sudo reboot
+```
+
+This will enable SP0 where those are the pins which will be seleted, only Software Chip Select is activated with the default pins:
```

Comment:
Thanks for adding this. Cost me a few hours to find out why hardware CS wasn't working...

Maybe add a sentence about the difference between hw and software cs? This isn't clear from this text.

Reply:
Seems that now there is no difference, once set it is by default activated. The hardware dtoverlay is depreciated.

Reply:
I mean "software" as in "do it yourself". I've seen quite a few examples (don't remember where, though), where the CS pin is manually controlled directly in the application.

Reply:
Actually, just using `dtparam=spi=on` will not activate _any_ CS pin, you need to do it yourself (by manyually controlling the pin that is connected to the CS input line of the attached chip). My finding with adding `dtoverlay=spi0-hw-cs` will change the ALT setting to the one described in the manual and use that pin. 
So with your suggestion, we have actually three different possibilities for the CS pin: None, hardware and "driver controlled gpio". 
Can you check what "gpio readall" reports for i.e. pin 27 when your line `dtoverlay=spi0-2cs,cs0_pin=27,cs1_pin=22` is used?

Reply:
My understanding is that now you choose the pins you want, reason why the `dtoverlay)spi0-hw-cs` is deprecated. Also `gpio` is not available anymore. See the note from the WiringPi author: http://wiringpi.com/wiringpi-deprecated/
It seems, reading the doc, that `dtparam=spio=on`will activate both default pins as Chip Select. Reason why I linked to the latest doc. Now, for previous versions, it seems it was not the case.
If you have any working alternative for `gpio readall`, happy to try!

Reply:
Well, at least the spi0-hw-cs works... I'll need to redo my SPL tests when I find time - last time I tried (when I wrote that issue) `dtparam=spi=on` would definitelly not activate any CS pins. 

The latest version of WiringPi, 2.52, does work on the Pi4 (and is very helpful in analyzing pin multiplexing behavior). However you need to download it from the website, it's not available via apt. 

With #1128 in place, we would be able to extend the `DeviceApiTester` tool to completelly replace `gpio`.

Reply:
I fully support a board abstract. We discussed it with @krwq. With all the request to support various baords, we need anyway a HAL/PAL approach.
I can try to clone the #1128 and see what I get!